### DISCUSSION TITLE
Severe sim lag

### DISCUSSION THREAD

**Discussion by JustLuckyRobotics** [cyberbotics/wrestling#61] [General]
We have been testing during our fights the actual sim time that is processed, when requesting the default 20ms time step our robot skips about 15-20% of the frames on average with the worst offender being over 40%. As in the supplied time step is 40ms instead of the requested 20. in some fights, by the time 25 seconds have passed 500 frames have been skipped. we have performance timed our code and it runs well below 1ms so is not that our code is over running.

We found this causes significant issues using multiple libraries supplied, for example the pose estimator is essentially useless under these conditions. the pose estimation algorithms are sensitive both to the uniformity of the time step and also to the update frequency. 50Hz is borderline fast enough to get a useable pose estimation as it is (most imu pose estimation algorithms really want 100hz+), dropping the update down to an average of 30fps, which we are regularly seeing, makes it useless. 

This issue effects multiple other libraries, some issues CAN be worked around however this issue appears to be getting worse as people develop more sophisticated systems. We have had to combat multiple bugs which caused our robot the lose entirely down to this lag issue.  As we do not have access to the other robots time step, we have no idea whether both bots are suffering an equal amount or if the issue is giving the participant or opponent an unfair advantage depending on which gets called first in the hierarchy. 

While the lag appears to be prevalent throughout the fights, the issue is worsened when the models are clipping into each other, with the greatest issue being when the robot intersects with it self (eg arms clipping into the hips)

A couple potential thoughts on how to mitigate the issue:
-turning off fast mode
    unsure if this will help but considering is appears the sim is unable to keep up with the actions within slowing the calculation speed 
    might help. this would increase the evaluation time, so those using auto commit bots might have to be asked to kindly stop doing 
    so or limit daily games to ensure a fair comp for all
-Enable the synchronisation within the world file
    I'm unsure however if this has the effect of synchronising the controllers of both bots it would at least make the environment fair 
    for both parties if no other mitigation or solutions can be found
-Change the joint angle limits to prevent self clipping

this might be a inherent limitation of the setup/software for the competition, however we thought it worth mentioning in case it reveals an underlying bug which can help improve the software and the comp as a whole.

**Comment by omichel** [comment]
Thank you for reporting this problem.

However, I believe your suggestions to improve the situation won't work:
- turning off the fast mode will increase the CPU and GPU consumption of Webots, so it would probably degrade the situation instead of improving it.
- enabling the synchronization within the world file would actually fix the problem. However, it would would open a new problem: a robot could block the simulation for an unlimited amount of time or it could slow down the simulation significantly, which we want to avoid: robot controllers should be able to run in real-time otherwise, this is not realistic and would make the simulation runs last too long.
- changing the joint angle limits to prevent self-clipping may reduce the CPU consumption of Webots, however I believe those limits are already correctly calibrated against the real robot. If you believe some limits could be adjusted to reduce the self-collisions, please let me know specifically which ones. 

Each robot controller is running inside docker container with a limited number of virtual CPU cores, e.g., 3 virtual CPU cores as you can see [here](https://github.com/cyberbotics/wrestling/blob/main/webots.yml#L9) and [there](https://github.com/cyberbotics/competition-record-action/blob/main/metascript/animation.py#L199-L200).

Webots is currently not limited in its use of CPU core, as we believed that it wouldn't use much more than 2, which should be fine since the self-hosted machine has a total of 8 virtual cores, e.g., 3 are allocated to the participant controller, 3 are allocated to the opponent controller and 2 are allocated to Webots.

However, we have no guarantee that Webots is using only 2 threads and if the simulation gets heavy (due to a large number of collisions), it could be that Webots starts using more thus affecting the CPU resources of the controller programs.

Therefore I believe that limiting explicitly the CPU cores allocated to the controllers and to Webots could help improving the situation by using:
```
--cpuset-cpus="0-1" # for webots
--cpuset-cpus="2-4" # for the participant
--cpuset-cpus="5-7" # for the opponent
```

I will implement this solution and let you test it (I will post a new message here when it is implemented and ready for testing).

**Reply by JustLuckyRobotics** [reply]
thank you for explaining why my suggestions will not work, the change of joint limits would have to be implemented similar to the head pitch/yaw joints where the angles are limited depending on the current position of other joints so would likely be far too complicated to implement. Synchronisation sounds like the ideal solution if there was the option to have webots ignore a controller if it takes longer than the specified time_step,  a bit like a watchdog check, without such an option you are correct in that it opens up avenues for abuse.

your suggested solution sounds best and interested to test it out. We will have to wait until someone challenges us as we are currently #1 and unable to trigger a fight unless you are capable of triggering a rerun of #2 v #1. however i doubt it wont be long until we are challenged every 6 minutes again.

**Reply by omichel** [reply]
I will manually trigger a new game of SugarSkull versus Chrake Peith once my patch is ready and let you know about it here.

**Reply by omichel** [reply]
This was implemented in https://github.com/cyberbotics/competition-record-action/pull/27 and is now in production. Please let me know if you see any better behavior of your controller. See [last game logs](https://github.com/cyberbotics/wrestling/actions/runs/4480949081/jobs/7876951709).

**Reply by JustLuckyRobotics** [reply]
no positive change, potentially worsened the issue. last few fights been dropping 38% of frames.

**Reply by omichel** [reply]
How do you see in your logs that your controller has been dropping 38% of the frames? It seems that your controller program is pretty computer intensive, utilizing up to 3 CPU cores at full load, thus, this is normal that your controller may skip some frames. A possible solution to ensure you don't skip any frame would be to implement multi-threading in your controller: one thread could simply loop on the `Robot.step()` function while another thread would perform heavy computations. This way, you shouldn't skip any frame. However, if your heavy compute thread is too long to performs its calculus, the outcome of the calculus may arrive too late for the next step...

> we have performance timed our code and it runs well below 1ms so is not that our code is over running.

Are you sure? It looks like your code is relying a lot on image processing and other compute intensive algorithms and I would be surprised it runs under 1ms... How did you time your code?

I have an idea to gain some compute time for controller programs. One of the problem is that during the `Robot.step(20)` function, the controller is idle for about 20 ms, waiting for the green light of Webots. During this time, Webots compute the physics according to the latest motor command and computes the sensor values and if there are still time, it waits to respect the total real time duration of the step, e.g., 20 real milliseconds. To use this idle time to perform computations, you can either rely on multi-threading in your controller code (as suggested above) or use the `Robot.stepBegin(20)` and `Robot.stepEnd()` instead of `Robot.step(20)`. Unlike `Robot.step(20)` which will cause your controller program to wait until Webots complete its simulation step, the `Robot.stepBegin(20)` and `Robot.stepEnd()` methods will return immediately, thus you should be able to use the time (about 20ms, since Webots is running in real time) in between these calls to make some heavy computation. Note however that these computations will obviously rely on the previous sensor measurements. See the documentation of these functions [here](https://cyberbotics.com/doc/reference/robot?tab-language=python#wb_robot_step_begin). Please let me know if that helps.

**Reply by JustLuckyRobotics** [reply]
We have already implemented this last night and it indeed solved our issues thank you. runs great now.

As should be standard, to ensure our code is performing as intended we use pythons time.perf at multiple points to both gauge overall code duration and specific function runtime while developing and debugging our code. we have been doing this since the beginning on our local sim (averaging <1ms with 6ms blips) and is what lead us to investigate the discrepancies we were seeing in the evaluation server.
